pkg/releaselock/rpc.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rpc
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"net/rpc"
+	"strconv"
+	"time"
+
+	"github.com/prashanthpai/sunrpc"
+	"k8s.io/klog/v2"
+)
+
+const (
+	inbandLockReleaseProgramNumber   = uint32(200002)
+	inbandLockReleaseProgramVersion  = uint32(1)
+	inbandLockReleaseProcedureNumber = uint32(1)
+	inbandLockReleaseProcedureName   = "IN_BAND_PROPRIETARY_LOCK_OPS_PROG.RELEASE_ALL_LOCKS"
+
+	pmapProgramNumber  = uint32(100021)
+	pmapProgramVersion = uint32(4)
+	pmapPort           = "111"
+
+	protocol = "tcp"
+)
+
+type releaseLockResponse struct {
+	status releaseLockStatus
+}
+
+type releaseLockStatus uint32
+
+// ReleaseLock calls the Filestore server to remove all advisory locks for a given GKE node IP.
+// hostIP is the internal IP address of the Filestore instance.
+// clientIP is the internal IP address of the GKE node.
+func ReleaseLock(hostIP, clientIP string) error {
+	procedureID := sunrpc.ProcedureID{
+		ProgramNumber:   inbandLockReleaseProgramNumber,
+		ProgramVersion:  inbandLockReleaseProgramVersion,
+		ProcedureNumber: inbandLockReleaseProcedureNumber,
+	}
+	procedure := sunrpc.Procedure{
+		ID:   procedureID,
+		Name: inbandLockReleaseProcedureName,
+	}
+	// TODO(tyuchn): move the registration out to driver initialization.
+	if err := sunrpc.RegisterProcedure(procedure, true /* validateProcName */); err != nil {
+		return fmt.Errorf("failed to register procedure %+v: %w", procedure, err)
+	}
+
+	// Get port from portmapper.
+	hostAddress := fmt.Sprintf("%s:%s", hostIP, pmapPort)
+	port, err := sunrpc.PmapGetPort(hostAddress, pmapProgramNumber, pmapProgramVersion, sunrpc.IPProtoTCP)
```

Comment:
Can you put some debug logs around the various steps like portmapper handshake, dial, RPC, and response

Reply:
Done